sui_programmability/examples/nft_standard/sources/example_marketplace.move

```diff
@@ -0,0 +1,34 @@
+module nft_standard::example_marketplace {
+
+    /// A shared object holding NFT listings
+    struct Marketplace {
+        /// NFT's listed for sale in this marketplace, indexed by their id's
+        // note: Table and ObjectTable will be added by https://github.com/MystenLabs/sui/issues/4203
+        listings: Table<ID, Listing>,
+        // commission taken on each filled listing. a flat fee, for simplicity.
+        commission: u64
+    }
+
+    struct Listing has store {
+        /// Price of the item in SUI
+        price: u64,
+        /// Capability to pull the item out of the appropriate `Safe` when making a sale
+        transfer_cap: TransferCap
+    }
+
+    fun init() {
+        // share the marketplace object, set the initial commission
+    }
+
+    public fun list(transfer_cap: TransferCap, marketplace: &mut Marketplace) {
+        // create a listing from transfer_cap add it to marketplace
+    }
+
+    public fun buy<T>(
+        royalty: RoyaltyReceipt<T>, coin: &mut Coin<SUI>, id: ID, safe: &mut Safe<T>, marketplace: &mut Marketplace
```

Comment:
I reckon the trading contracts should be the ones enforcing royalties, otherwise, as a client, I would write my own interaction with this method in which I fool the `buy_nft` fn to get a `RoyaltyReceipt` for cheap.

In other cases, I might not know the price I will pay upfront. If royalty is a percentage cut, and I buy something from an orderbook, the state can change since I sent the request.